kore/src/Kore/Step/Simplification/OverloadSimplifier.hs

```diff
@@ -0,0 +1,111 @@
+{- |
+Copyright   : (c) Runtime Verification, 2020
+License     : NCSA
+
+-}
+module Kore.Step.Simplification.OverloadSimplifier
+    ( OverloadSimplifier (..)
+    , mkOverloadSimplifier
+    , injectTermTo
+    ) where
+
+import Control.Exception
+    ( assert
+    )
+import qualified Data.Functor.Foldable as Recursive
+import Data.Maybe
+    ( listToMaybe
+    )
+import qualified GHC.Stack as GHC
+
+import Kore.Attribute.Synthetic
+    ( synthesize
+    )
+import Kore.IndexedModule.OverloadGraph
+    ( OverloadGraph
+    )
+import qualified Kore.IndexedModule.OverloadGraph as OverloadGraph
+import Kore.IndexedModule.SortGraph
+    ( SortGraph
+    , isSubsortOf
+    )
+import Kore.Internal.ApplicationSorts
+    ( applicationSortsResult
+    )
+import qualified Kore.Internal.Symbol as Symbol
+import Kore.Internal.TermLike
+
+{- | Data structure encapsulating functionality for handling overloaded
+symbols during unification.
+-}
+data OverloadSimplifier =
+    OverloadSimplifier
+        { isOverloading :: Symbol -> Symbol -> Bool
+        -- ^ Whether the first argument is overloading the second
+        , isOverloaded :: Symbol -> Bool
+        -- ^ Whether the symbol is overloaded
+        , resolveOverloading
+            :: forall variable
+            .  GHC.HasCallStack
+            => InternalVariable variable
+            => Inj ()
+            -> Symbol
+            -> [TermLike variable]
+            -> TermLike variable
+        -- ^ Apply an overloading equation
```

Comment:
Could you expand a bit? I think that this injects the list of TermLike so that they fit the symbol's arguments. If so, could you add this to the comment? If not, could you explain what happens?